core/amber/src/main/scala/edu/uci/ics/texera/web/resource/aiassistant/python_abstract_syntax_tree.py

```diff
@@ -0,0 +1,39 @@
+import ast
+import json
+import sys
+import base64
+
+class TypeAnnotationVisitor(ast.NodeVisitor):
+    def __init__(self, start_line_offset=0):
+        self.untyped_args = []
+        # To calculate the correct start line
+        self.start_line_offset = start_line_offset
+
+    def visit_FunctionDef(self, node):
+        for arg in node.args.args:
+            # Self is not an argument
+            if arg.arg == 'self':
+                continue
+            if not arg.annotation:
+                # +1 and -1 is to change ast line/col to monaco editor line/col for the range
+                start_line = arg.lineno + self.start_line_offset - 1
+                start_col = arg.col_offset + 1
+                end_line = start_line
+                end_col = start_col + len(arg.arg)
+                self.untyped_args.append([arg.arg, start_line, start_col, end_line, end_col])
+        self.generic_visit(node)
+
+def find_untyped_variables(source_code, start_line):
+    tree = ast.parse(source_code)
+    visitor = TypeAnnotationVisitor(start_line_offset=start_line)
+    visitor.visit(tree)
+    return visitor.untyped_args
+
+if __name__ == "__main__":
+    # First argument is the code
+    encoded_code = sys.argv[1]
+    # Second argument is the start line
+    start_line = int(sys.argv[2])
```

Comment:
I don't understand the point of having a `start_line`. do you only consider part of user's code starting from a line? if so I think it is incorrect. let your module analyze the entire module provided by the user.

Reply:
Wouldn't it be more flexible to allow users to choose the range where they want to add type annotations? They can also choose to add annotations to all the code at once.

Reply:
I see. But the code would have dependencies, and may not always to be sequentially arranged in text. let's think of one case:

```
1 def g(x: int) -> int:
2     return int(x)
3 
4 def f(a):
5    return g(a)
```

if only selecting line 4, what would your annotation for `a` be? selecting a partial code starting from a line may not capture all the definitions, for example it may omit `g`'s definition in this case. Do you think that's a valid use case of your feature?

Reply:
Regardless of which lines the user selects for 'add all type annotations', the entire code within the Python UDF will be provided to the LLM. This ensures that the LLM has full context of the code for accurate annotation.

Reply:
Ok got it.